src/framework/audio/internal/audioengine.cpp

```diff
@@ -19,289 +19,147 @@
 
 #include "audioengine.h"
 
-#include <soloud.h>
-
 #include "log.h"
 #include "ptrutils.h"
-#include "../audioerrors.h"
+#include "audioerrors.h"
 
 using namespace mu::audio;
-using namespace mu::audio;
-
-struct AudioEngine::SL {
-    SoLoud::Soloud engine;
-};
 
 AudioEngine::AudioEngine()
 {
-    m_sl = std::shared_ptr<SL>(new SL);
-    m_sl->engine.mBackendData = this;
-}
-
-bool AudioEngine::isInited() const
-{
-    std::shared_ptr<IAudioDriver> drv = driver();
-    if (!drv) {
-        return false;
-    }
-
-    if (!drv->isOpened()) {
-        return false;
-    }
+    m_sequencer = std::make_shared<Sequencer>();
 
-    if (!m_inited) {
-        return false;
-    }
+    m_mixer = std::make_shared<Mixer>();
+    m_mixer->setClock(m_sequencer->clock());
 
-    return true;
-}
+    m_buffer = std::make_shared<AudioBuffer>();
+    m_buffer->setSource(m_mixer);
 
-mu::Ret AudioEngine::init()
-{
-    if (isInited()) {
-        return make_ret(Ret::Code::Ok);
-    }
-
-    int res = m_sl->engine.init(SoLoud::Soloud::CLIP_ROUNDOFF,
-                                SoLoud::Soloud::MUAUDIO,
-                                SoLoud::Soloud::AUTO,
-                                SAMPLE_GRANULARITY, // 1024
-                                2);
-
-    if (res == SoLoud::SO_NO_ERROR) {
-        LOGI() << "success inited audio engine";
-        m_inited = true;
-        m_initChanged.send(m_inited);
-        return make_ret(Ret::Code::Ok);
-    }
-
-    m_inited = false;
-    m_initChanged.send(m_inited);
-
-    Err err = Err::UnknownError;
-    if (SoLoud::INVALID_PARAMETER == res) {
-        err = Err::EngineInvalidParameter;
-    } else if (int(Err::DriverNotFound) == res) {
-        err = Err::DriverNotFound;
-    } else if (int(Err::DriverOpenFailed) == res) {
-        err = Err::DriverOpenFailed;
-    }
-
-    LOGE() << "failed inited audio engine, err: " << int(err);
-    return make_ret(err);
-}
-
-void AudioEngine::deinit()
-{
-    m_sl->engine.deinit();
-    m_inited = false;
-    m_initChanged.send(m_inited);
-}
-
-mu::async::Channel<bool> AudioEngine::initChanged() const
-{
-    return m_initChanged;
-}
+    m_sequencer->audioTrackAdded().onReceive(this, [this](ISequencer::audio_track_t player) {
+        m_mixer->addChannel(player);
+    });
 
-float AudioEngine::sampleRate() const
-{
-    return m_sl->engine.getBackendSamplerate();
+    m_worker = std::make_shared<AudioThread>();
```

Comment:
This is a design problem.
AudioEngine generic, platform-independent class (sort of)
The presence or absence of AudioThread depends on the platform and implementation. For example, there is no AudioThread for working with Web Worker or a separate process.
So, AudioThread cannot be part of AudioEngine, it must (if any) be above AudioEngine

Reply:
fixed  https://github.com/musescore/MuseScore/pull/7266